### DISCUSSION TITLE
Scaling / Parallelism / Concurrency safe?

### DISCUSSION THREAD

**Discussion by Manouchehri** [langfuse/langfuse#1016] [Support] [answered]
Is Langfuse expected to work properly if I scale the same Langfuse image to multiple containers, that all connect to the same PostgreSQL endpoint?

(We have Langfuse deployed to Google Cloud Run, so it automatically scales up and down for us. I can cap it to only 1 concurrent container if needed though.)

**Comment by marcklingen** [comment, marked as answer]
This should work as expected as long as you upgrade all containers roughly at the same time. Running multiple instances on different application versions can lead to:
- mismatch of application <> database schema (migrations are applied by the application container during start) -> application/api errors
- mismatch of frontend <> frontend api -> unstable frontend

**Reply by Manouchehri** [reply]
Thanks!

If we have an outdated container that's left lingering for a few seconds after a new Langfuse version has started and already migrated the database schema, will the old outdated container wreck anything?

**Reply by marcklingen** [reply]
I think the risk of wrecking something is low, but you might see some API/UI errors during this time. All client SDKs automatically retry on errors, effectively bridging the short timeframe.

**Reply by Manouchehri** [reply]
Sweet, shouldn't be a concern then, thanks! As long as `/api/public/health` works as expected (returning a non-200) in the old container when it encounters database errors, Google Cloud Run should kill the old container(s) pretty quickly. (I should probably knock down the failure threshold..)

<img width="502" alt="image" src="https://github.com/langfuse/langfuse/assets/7232674/a29ad0a3-f0fb-42a1-a571-12f50962038b">

**Reply by marcklingen** [reply]
That's not how the health endpoint works currently, it only checks if the database is available, not if the schema is out of sync with the application ([source](https://github.com/langfuse/langfuse/blob/main/src/pages/api/public/health.ts))

**Reply by Manouchehri** [reply]
Hmm, just throwing ideas out there, maybe that endpoint could check if the database version has changed since startup time? (Not sure if that would slow things down though.)

This isn't a huge deal as it's a theoretical problem that I don't know if we will ever hit. =)

**Reply by marcklingen** [reply]
Agree, had the exact same though of "should be possible" and "let's see whether it is actually an issue". Let me know in case this does not work smoothly
